File: packages/sdk/kotlin/admin/src/main/kotlin/dev/edgebase/sdk/admin/AdminPushClient.kt
// EdgeBase Kotlin SDK — Admin push notification client.
//
// Admin-side push notification management.
// Requires Service Key.

package dev.edgebase.sdk.admin

import dev.edgebase.sdk.admin.generated.GeneratedAdminApi
import dev.edgebase.sdk.core.HttpClient

/** Push payload. */
data class PushPayload(
    val title: String? = null,
    val body: String? = null,
    val image: String? = null,
    val sound: String? = null,
    val badge: Int? = null,
    val data: Map<String, Any>? = null,
    val silent: Boolean? = null,
    val collapseId: String? = null,
    val ttl: Int? = null,
    val apns: Map<String, Any>? = null,
    val fcm: Map<String, Any>? = null,
    val wns: Map<String, Any>? = null,
    val web: Map<String, Any>? = null
) {
    /** Convert to Map for JSON serialization (anyToJsonElement handles Map, not data class). */
    fun toMap(): Map<String, Any?> {
        val map = mutableMapOf<String, Any?>()
        if (title != null) map["title"] = title
        if (body != null) map["body"] = body
        if (image != null) map["image"] = image
        if (sound != null) map["sound"] = sound
        if (badge != null) map["badge"] = badge
        if (data != null) map["data"] = data
        if (silent != null) map["silent"] = silent
        if (collapseId != null) map["collapseId"] = collapseId
        if (ttl != null) map["ttl"] = ttl
        if (apns != null) map["apns"] = apns
        if (fcm != null) map["fcm"] = fcm
        if (wns != null) map["wns"] = wns
        if (web != null) map["web"] = web
        return map
    }
}

/** Result of a send operation. */
data class PushSendResult(
    val sent: Int,
    val failed: Int,
    val removed: Int
)

/** Device token record (Admin only). */
data class DeviceTokenRecord(
    val deviceId: String,
    val token: String,
    val platform: String,
    val updatedAt: String,
    val deviceInfo: Map<String, String>? = null,
    val metadata: Map<String, Any>? = null
)

/** Send log entry. */
data class PushLogEntry(
    val sentAt: String,
    val userId: String,
    val platform: String,
    val status: String,
    val collapseId: String? = null,
    val error: String? = null
)

/**
 * Server-side push notification management (Admin).
 * Enables sending messages, viewing tokens, and reading logs.
 * Requires a Service Key with appropriate permissions.
 */
class AdminPushClient(private val client: HttpClient) {

    private val adminCore = GeneratedAdminApi(client)

    private fun asNullableAnyMap(value: Any?): Map<String, Any?>? {
        val map = value as? Map<*, *> ?: return null
        return buildMap {
            for ((key, entryValue) in map) {
                if (key is String) {
                    put(key, entryValue)
                }
            }
        }
    }

    private fun asAnyMap(value: Any?): Map<String, Any>? {
        val map = value as? Map<*, *> ?: return null
        return buildMap {
            for ((key, entryValue) in map) {
                if (key is String && entryValue != null) {
                    put(key, entryValue)
                }
            }
        }
    }

    private fun asStringMap(value: Any?): Map<String, String>? {
        val map = value as? Map<*, *> ?: return null
        return buildMap {
            for ((key, entryValue) in map) {
                if (key is String && entryValue is String) {
                    put(key, entryValue)
                }
            }
        }
    }

    private fun asMapList(value: Any?): List<Map<String, Any?>> {
        val list = value as? List<*> ?: return emptyList()
        return list.mapNotNull(::asNullableAnyMap)
    }

    /**
     * Send a push notification to a specific user.
     * The server will route it to all the user's registered devices.
     */
    suspend fun send(userId: String, payload: PushPayload): PushSendResult {
        val body = mapOf("userId" to userId, "payload" to payload.toMap())
        @Suppress("UNCHECKED_CAST")
        val res = client.post("/push/send", body) as Map<String, Any?>
        return PushSendResult(
            sent = (res["sent"] as? Number)?.toInt() ?: 0,
            failed = (res["failed"] as? Number)?.toInt() ?: 0,
            removed = (res["removed"] as? Number)?.toInt() ?: 0
        )
    }

    /**
     * Send a push notification to multiple users.
     * Batched sequentially in chunks of 500 on the server.
     */
    suspend fun sendMany(userIds: List<String>, payload: PushPayload): PushSendResult {
        val body = mapOf("userIds" to userIds, "payload" to payload.toMap())
        @Suppress("UNCHECKED_CAST")
        val res = client.post("/push/send-many", body) as Map<String, Any?>
        return PushSendResult(
            sent = (res["sent"] as? Number)?.toInt() ?: 0,
            failed = (res["failed"] as? Number)?.toInt() ?: 0,
            removed = (res["removed"] as? Number)?.toInt() ?: 0
        )
    }

    /**
     * Send a push notification to a specific device token.
     */
    suspend fun sendToToken(token: String, payload: PushPayload, platform: String? = null): PushSendResult {
        val body = mapOf("token" to token, "payload" to payload.toMap(), "platform" to (platform ?: "web"))
        @Suppress("UNCHECKED_CAST")
        val res = client.post("/push/send-to-token", body) as Map<String, Any?>
        return PushSendResult(
            sent = (res["sent"] as? Number)?.toInt() ?: 0,
            failed = (res["failed"] as? Number)?.toInt() ?: 0,
            removed = (res["removed"] as? Number)?.toInt() ?: 0
        )
    }

    /**
     * View registered devices for a user.
     * The token value itself is returned because it's an Admin API.
     */
    suspend fun getTokens(userId: String): List<DeviceTokenRecord> {
        @Suppress("UNCHECKED_CAST")
        val res = adminCore.getPushTokens(mapOf("userId" to userId)) as Map<String, Any?>
        val items = asMapList(res["items"])
        return items.map { item ->
            DeviceTokenRecord(
                deviceId = item["deviceId"] as String,
                token = item["token"] as String,
                platform = item["platform"] as String,
                updatedAt = item["updatedAt"] as String,
                deviceInfo = asStringMap(item["deviceInfo"]),
                metadata = asAnyMap(item["metadata"])
            )
        }
    }

    /**
     * Get recent push logs for a user (last 24h).
     */
    suspend fun getLogs(userId: String, limit: Int = 50): List<PushLogEntry> {
        val query = mutableMapOf("userId" to userId)
        if (limit > 0) query["limit"] = limit.toString()
        @Suppress("UNCHECKED_CAST")
        val res = adminCore.getPushLogs(query) as Map<String, Any?>
        val items = asMapList(res["items"])
        return items.map { item ->
            PushLogEntry(
                sentAt = item["sentAt"] as String,
                userId = item["userId"] as String,
                platform = item["platform"] as String,
                status = item["status"] as String,
                collapseId = item["collapseId"] as? String,
                error = item["error"] as? String
            )
        }
    }

    /**
     * Send a push notification to an FCM topic.
     */
    @Suppress("UNCHECKED_CAST")
    suspend fun sendToTopic(topic: String, payload: Map<String, Any?>): Map<String, Any?> {
        val res = client.post("/push/send-to-topic", mapOf("topic" to topic, "payload" to payload))
        return res as? Map<String, Any?> ?: emptyMap()
    }

    /**
     * Broadcast a push notification to all devices via /topics/all.
     */
    @Suppress("UNCHECKED_CAST")
    suspend fun broadcast(payload: Map<String, Any?>): Map<String, Any?> {
        val res = client.post("/push/broadcast", mapOf("payload" to payload))
        return res as? Map<String, Any?> ?: emptyMap()
    }
}

File: packages/sdk/csharp/packages/admin/AdminPushClient.cs
using System;
using System.Collections.Generic;
using System.Threading;
using System.Threading.Tasks;
using EdgeBase.Generated;
// EdgeBase C# Admin SDK — Admin Push Client
// 서버사이드 푸시 알림 발송: send, sendMany, sendToToken, getTokens, getLogs.
// All HTTP calls delegate to Generated Admin Core (GeneratedAdminApi).
// No hardcoded API paths — the core is the single source of truth.

namespace EdgeBase.Admin
{

/// <summary>서버사이드 푸시 알림 발송 클라이언트.</summary>
public sealed class AdminPushClient
{
    private readonly JbHttpClient _http;
    private readonly GeneratedAdminApi _adminCore;

    internal AdminPushClient(JbHttpClient http)
    {
        _http = http;
        _adminCore = new GeneratedAdminApi(http);
    }

    /// <summary>특정 유저의 모든 디바이스에 푸시 발송.</summary>
    /// <returns>발송 결과 (sent, failed, removed)</returns>
    public async Task<PushResult> SendAsync(
        string userId, Dictionary<string, object?> payload,
        CancellationToken ct = default)
    {
        var body = new Dictionary<string, object?>
        {
            ["userId"]  = userId,
            ["payload"] = payload
        };
        var result = await _adminCore.PushSendAsync(body, ct);
        return PushResult.FromDict(result);
    }

    /// <summary>여러 유저에게 대량 발송.</summary>
    public async Task<PushResult> SendManyAsync(
        string[] userIds, Dictionary<string, object?> payload,
        CancellationToken ct = default)
    {
        var body = new Dictionary<string, object?>
        {
            ["userIds"] = userIds,
            ["payload"] = payload
        };
        var result = await _adminCore.PushSendManyAsync(body, ct);
        return PushResult.FromDict(result);
    }

    /// <summary>특정 디바이스 토큰에 발송.</summary>
    public async Task<PushResult> SendToTokenAsync(
        string token, Dictionary<string, object?> payload,
        string? platform = null, CancellationToken ct = default)
    {
        var body = new Dictionary<string, object?>
        {
            ["token"]    = token,
            ["payload"]  = payload,
            ["platform"] = platform ?? "web"
        };
        var result = await _adminCore.PushSendToTokenAsync(body, ct);
        return PushResult.FromDict(result);
    }

    /// <summary>유저의 등록된 디바이스 목록 조회 (토큰 값 미포함).</summary>
    public async Task<List<Dictionary<string, object?>>> GetTokensAsync(
        string userId, CancellationToken ct = default)
    {
        var query = new Dictionary<string, string> { ["userId"] = userId };
        var result = await _adminCore.GetPushTokensAsync(query, ct);
        var tokens = JsonHelper.ExtractList(result, "tokens");
        return tokens.Count > 0 ? tokens : JsonHelper.ExtractList(result, "items");
    }

    /// <summary>발송 로그 조회 (24시간 TTL).</summary>
    public async Task<List<Dictionary<string, object?>>> GetLogsAsync(
        string userId, int? limit = null, CancellationToken ct = default)
    {
        var query = new Dictionary<string, string> { ["userId"] = userId };
        if (limit.HasValue) query["limit"] = limit.Value.ToString();
        var result = await _adminCore.GetPushLogsAsync(query, ct);
        var logs = JsonHelper.ExtractList(result, "logs");
        return logs.Count > 0 ? logs : JsonHelper.ExtractList(result, "items");
    }

    /// <summary>FCM 토픽에 푸시 발송.</summary>
    public async Task<Dictionary<string, object?>> SendToTopicAsync(
        string topic, Dictionary<string, object?> payload,
        CancellationToken ct = default)
    {
        var body = new Dictionary<string, object?>
        {
            ["topic"]   = topic,
            ["payload"] = payload
        };
        return await _adminCore.PushSendToTopicAsync(body, ct);
    }

    /// <summary>전체 디바이스에 브로드캐스트.</summary>
    public async Task<Dictionary<string, object?>> BroadcastAsync(
        Dictionary<string, object?> payload,
        CancellationToken ct = default)
    {
        var body = new Dictionary<string, object?>
        {
            ["payload"] = payload
        };
        return await _adminCore.PushBroadcastAsync(body, ct);
    }
}

/// <summary>푸시 발송 결과.</summary>
public sealed class PushResult
{
    /// <summary>성공적으로 발송된 수.</summary>
    public int Sent { get; }
    /// <summary>발송 실패 수.</summary>
    public int Failed { get; }
    /// <summary>만료/삭제된 토큰 수 (자동 정리됨).</summary>
    public int Removed { get; }

    public PushResult(int sent, int failed, int removed)
    {
        Sent    = sent;
        Failed  = failed;
        Removed = removed;
    }

    internal static PushResult FromDict(Dictionary<string, object?> dict)
    {
        return new PushResult(
            JsonHelper.ExtractInt(dict, "sent"),
            JsonHelper.ExtractInt(dict, "failed"),
            JsonHelper.ExtractInt(dict, "removed"));
    }
}
}
